plugins/wasi_nn/ggml.cpp

```diff
@@ -366,100 +368,126 @@ Expect<ErrNo> compute(WasiNNEnvironment &Env, uint32_t ContextId) noexcept {
-  llama_batch LlamaBatch = llama_batch_init(GraphRef.BatchSize, 0);
-  LlamaBatch.n_tokens = CxtRef.LlamaInputs.size();
-  for (int32_t I = 0; I < LlamaBatch.n_tokens; I++) {
-    LlamaBatch.token[I] = CxtRef.LlamaInputs[I];
-    LlamaBatch.pos[I] = I;
-    LlamaBatch.seq_id[I] = 0;
-    LlamaBatch.logits[I] = false;
-  }
-
-  // llama_decode will output logits only for the last token of the prompt
-  LlamaBatch.logits[LlamaBatch.n_tokens - 1] = true;
-  if (llama_decode(CxtRef.LlamaContext, LlamaBatch) != 0) {
-    spdlog::info("[WASI-NN] GGML backend: llama_decode() failed"sv);
-    return ErrNo::RuntimeError;
-  }
-
-  int NCur = LlamaBatch.n_tokens;
-  while (NCur < MaxContextSize && NCur < NPredict) {
-    // Sample the next token
-    auto NVocab = llama_n_vocab(GraphRef.LlamaModel);
-    auto *Logits =
-        llama_get_logits_ith(CxtRef.LlamaContext, LlamaBatch.n_tokens - 1);
-
-    std::vector<llama_token_data> Candidates;
-    Candidates.reserve(NVocab);
-    for (llama_token TokenId = 0; TokenId < NVocab; TokenId++) {
-      Candidates.emplace_back(llama_token_data{TokenId, Logits[TokenId], 0.0f});
-    }
-    llama_token_data_array CandidatesP = {Candidates.data(), Candidates.size(),
-                                          false};
-
-    // Sample the most likely token
-    const llama_token NewTokenId =
-        llama_sample_token_greedy(CxtRef.LlamaContext, &CandidatesP);
-
-    // Is it an end of stream?
-    if (NewTokenId == llama_token_eos(CxtRef.LlamaContext) ||
-        NCur == MaxContextSize || NCur == NPredict) {
-      break;
-    }
-
-    std::string NextToken =
-        llama_token_to_piece(CxtRef.LlamaContext, NewTokenId);
-
-    // When setting StreamStdout, we print the output to stdout.
-    if (GraphRef.StreamStdout) {
-      std::cout << NextToken << std::flush;
-    }
-
-    // Append the new token.
-    CxtRef.LlamaOutputs += NextToken;
-
-    // Prepare the next batch
-    LlamaBatch.n_tokens = 0;
+  gpt_params GPTParams;
+  struct llama_sampling_context *CtxSampling =
+      llama_sampling_init(GPTParams.sparams);
+  std::vector<llama_token> Embd;
+  int NPast = 0;
+  int NConsumed = 0;
+  int NRemain = GraphRef.NPredict;
+  // Initialize the llama context.
+  llama_context_params ContextParams = llama_context_default_params();
+  ContextParams.n_ctx = GraphRef.CtxSize;
+  ContextParams.n_batch = GraphRef.BatchSize;
+  CxtRef.LlamaContext =
+      llama_new_context_with_model(GraphRef.LlamaModel, ContextParams);
+  int NCtx = llama_n_ctx(CxtRef.LlamaContext);
+  // Minus 4 for the special tokens. (Such as <BOS>, <EOS>, ... tokens.)
+  const int MaxTokensListSize = NCtx - 4;
+  // Use the const sequence id here.
+  const int SequenceId = 0;
+  while (NRemain >= 0) {
+    // Preidct
+    if (!Embd.empty()) {
+      // Input too long.
+      if (static_cast<int>(Embd.size()) > MaxTokensListSize) {
+        spdlog::error(
+            "[WASI-NN] GGML backend: Error: prompt too long ({} tokens, max {})"sv,
```

Comment:
Let's give a recommended solution, such as stripping the input to match the maximum token length(n_ctx).

Reply:
Just updated the error message for recommending reduce input tokens.